test/common/network/win32_socket_handle_impl_test.cc

```diff
@@ -0,0 +1,182 @@
+
+  absl::FixedArray<char> buf(10);
+  auto rc = io_handle_.recv(buf.data(), buf.size(), 0);
+  EXPECT_EQ(rc.return_value_, 10);
+}
+
+TEST_F(Win32SocketHandleImplTest, RecvWithPeekReactivatesReadOnBlock) {
+  Api::MockOsSysCalls os_sys_calls;
+  TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls(&os_sys_calls);
+  EXPECT_CALL(os_sys_calls, readv(_, _, _))
+      .Times(1)
+      .WillOnce(Return(Api::SysCallSizeResult{-1, SOCKET_ERROR_AGAIN}));
+
+  EXPECT_CALL(*file_event_, registerEventIfEmulatedEdge(_));
+  absl::FixedArray<char> buf(10);
+  auto rc = io_handle_.recv(buf.data(), buf.size(), MSG_PEEK);
+  EXPECT_EQ(rc.err_->getErrorCode(), Api::IoError::IoErrorCode::Again);
+}
+
+TEST_F(Win32SocketHandleImplTest, RecvWithPeekFlagReturnsFinalError) {
+  Api::MockOsSysCalls os_sys_calls;
+  TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls(&os_sys_calls);
+  constexpr int data_length = 10;
+  std::string data(data_length, '*');
+  absl::FixedArray<iovec> peek_iov(1);
+  peek_iov[0].iov_base = static_cast<void*>(data.data());
+  peek_iov[0].iov_len = data.length();
+  EXPECT_CALL(os_sys_calls, readv(_, _, _))
+      .Times(2)
+      .WillOnce(Invoke([&](os_fd_t, const iovec* iov, int num_iov) {
+        // Gcc treats the variables as unused and this causes
+        // a compilation failure.
+        UNREFERENCED_PARAMETER(iov);
+        UNREFERENCED_PARAMETER(num_iov);
+        iov = peek_iov.begin();
+        num_iov = 1;
+        return Api::SysCallSizeResult{data_length / 2, 0};
+      }))
+      .WillOnce(Return(Api::SysCallSizeResult{-1, SOCKET_ERROR_CONNRESET}));
+
+  absl::FixedArray<char> buf(data_length);
+  auto rc = io_handle_.recv(buf.data(), buf.size(), MSG_PEEK);
+  EXPECT_EQ(rc.err_->getErrorCode(), Api::IoError::IoErrorCode::ConnectionReset);
+}
+
+TEST_F(Win32SocketHandleImplTest, ReadvWithPeekShouldReadFromBuffer) {
+  Api::MockOsSysCalls os_sys_calls;
+  TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls(&os_sys_calls);
+  constexpr int data_length = 10;
+  std::string data(data_length, '*');
+  absl::FixedArray<iovec> peek_iov(1);
+  peek_iov[0].iov_base = static_cast<void*>(data.data());
+  peek_iov[0].iov_len = data.length();
+  EXPECT_CALL(os_sys_calls, readv(_, _, _))
+      .WillOnce(Invoke([&](os_fd_t, const iovec* iov, int num_iov) {
+        // Gcc treats the variables as unused and this causes
+        // a compilation failure.
+        UNREFERENCED_PARAMETER(iov);
+        UNREFERENCED_PARAMETER(num_iov);
+        iov = peek_iov.begin();
+        num_iov = 1;
+        return Api::SysCallSizeResult{data_length, 0};
+      }));
+
+  absl::FixedArray<char> buf(data_length);
+  auto rc = io_handle_.recv(buf.data(), buf.size(), MSG_PEEK);
+  EXPECT_EQ(rc.return_value_, data_length);
+  // Second call should not make a system call, it should
+  // read from memory.
+  rc = io_handle_.recv(buf.data(), buf.size(), MSG_PEEK);
+  EXPECT_EQ(rc.return_value_, data_length);
```

Comment:
Would it be possible to add some expectations about the contents of buf at the end of these peek operations?

Reply:
yea, let me add that.